plugins/sigma/src/parse.cpp

```diff
@@ -275,47 +270,144 @@ caf::expected<expression> parse_search_id(const data& yaml) {
           // should become /X$/.
           op = relational_operator::ni;
         } else if (*i == "base64") {
-          // TODO
-          return caf::make_error(ec::unimplemented, "base64 modifier not yet "
-                                                    "implemented");
+          auto encode = [](const data& x) -> caf::expected<data> {
+            if (const auto* str = caf::get_if<std::string>(&x))
+              return detail::base64::encode(*str);
+            return caf::make_error(ec::type_clash, //
+                                   "base64 only works with strings");
+          };
+          transforms.emplace_back(encode);
         } else if (*i == "base64offset") {
-          // TODO
-          return caf::make_error(ec::unimplemented, "base64offset modifier not "
-                                                    "yet implemented");
+          auto encode = [](const data& x) -> caf::expected<data> {
+            const auto* str = caf::get_if<std::string>(&x);
+            if (!str)
+              return caf::make_error(ec::type_clash, //
+                                     "base64offset only works with strings");
+            static constexpr std::array<size_t, 3> start = {{0, 2, 3}};
+            static constexpr std::array<size_t, 3> end = {{0, 3, 2}};
+            std::vector<std::string> xs(3);
+            for (size_t i = 0; i < 3; ++i) {
+              auto padded = std::string(i, ' ') + *str;
+              auto b64 = detail::base64::encode(padded);
+              auto len = b64.size() - end[(str->size() + i) % 3];
+              xs[i] = b64.substr(start[i], len - start[i]);
+            }
+            return list{xs[0], xs[1], xs[2]};
+          };
+          transforms.emplace_back(encode);
         } else if (*i == "utf16le" || *i == "wide") {
-          // TODO
-          return caf::make_error(ec::unimplemented, "utf16le/wide modifier not "
-                                                    "yet implemented");
+          return caf::make_error(ec::unimplemented, //
+                                 "utf16le/wide not yet implemented");
+          // FIXME: the attempt below doesn't work yet, but gives an idea on
+          // what needs to be done algorithmically.
+          auto convert = [](const data& x) -> caf::expected<data> {
+            const auto* str = caf::get_if<std::string>(&x);
+            if (!str)
+              return caf::make_error(ec::type_clash, //
+                                     "utf16le/wide only works with strings");
+            // Hand-roll conversion.
+            std::string result;
+            result.reserve(str->size() * 2);
+            for (auto c : *str) {
+              result.push_back(c);
+              result.push_back(0x00);
+            }
```

Comment:
@dominiklohmann here's a naive attempt to hand-roll support for UTF16 strings. Can this work?

(It's currently dead code, but we have a story for making this work.)

Reply:
That's most certainly not correct—you cannot map multi-byte (or rather, multi-codepoint) UTF8 strings onto UTF16 codepoints like this. I propose throwing this out right now and asking users to do this conversion externally. Since most of the world only supports UTF8 nowadays I don't think we need to implement this at all, ever, unless someone is paying us enough money to go down this rabbit hole or std::text exists in standard C++.

Reply:
> I propose throwing this out right now and asking users to do this conversion externally. 

This is not an option. It looks like my question was not very clear. Sigma has a few UTF modifiers (`utf16le`, `utf16be`, and `utf8`) that take ASCII strings and produce the corresponding UTF output. These need to be supported along other modifiers.

Since our raw data is just bytes, we _can_ perform this translation easily for UTF16, since it's just byte padding. UTF8 is a different story.

> I don't think we need to implement this at all

This is not correct. In order to support the Sigma rules out there, we need to implement these modifiers.

Reply:
You don't validate here that your input is ASCII. YAML supports UTF-8, and that's what's likely to be your input here. If you want to do the naïve conversion, then you must verify that assumption first.

Regarding the use case I'm still not convinced. Yes, we need it to claim "full Sigma support." But does anyone _actually use_ it?

Reply:
> Regarding the use case I'm still not convinced. Yes, we need it to claim "full Sigma support." But does anyone actually use it?

Fair question. I git-grepped the existing rules in the repo:

```
❯ git grep '|wide' | wc -l
       1   <-- unit test only
❯ git grep '|utf16be' | wc -l
       0
❯ git grep '|utf16le' | wc -l
       0
❯ git grep '|utf8' | wc -l
       0
```

In comparison (also low)

```
❯ git grep '|base64' | wc -l
       9
❯ git grep '|base64offset' | wc -l
       8
```

So your hunch was correct, there are not many (at least official) rules that use the codepoint conversion features. That's helpful wrt. prioritization of this feature.